### PR TITLE
AUTO-680 Autoscaling group scaling cloudwatch policies

### PR DESCRIPTION
**Pull request by GypsyBojangles**
* AUTO-680 Autoscaling group scaling cloudwatch policies
* AUTO-772 Remove stack title from name of resources
* AUTO-732 Allow block device mappings to populate with defaults
* AUTO-643 Separate YAML loading and validation
* AUTO-733 Block device mappings as an object
* Create "complex object field mapping" to generalise loading data from complex objects
* Yaml.py now populates complext objects, not stack.py

Build passing https://gaautobots.atlassian.net/builds/browse/AM-AUT61
Build passoing https://gaautobots.atlassian.net/builds/browse/AM-AST54

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,12 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [1.3.1] - 26/08/2016
```
Comment (ga-autobots) [inline CHANGELOG.md]: date

File: amazonia/classes/asg.py [hunk 2]
```diff
@@ -27,6 +28,8 @@ def __init__(self, title, template, network_config, load_balancers, asg_config):
         self.lc = None
         self.cd_app = None
         self.cd_deploygroup = None
+        self.cw_alarms = []
+        self.scaling_polcies = []
```
Comment (ga-autobots) [inline amazonia/classes/asg.py]: policies

File: amazonia/classes/asg.py [hunk 3]
```diff
@@ -17,7 +19,6 @@ def __init__(self, title, template, network_config, load_balancers, asg_config):
         :param network_config: object containing network related config
         :param asg_config: object containing asg related config
         :param load_balancers: list of load balancers to associate autoscaling group with
-        :param block_devices_config: List containing block device mappings
         """
         super(Asg, self).__init__(vpc=network_config.vpc, title=title, template=template)
```
Comment (ga-autobots) [inline amazonia/classes/asg.py]: add comments

File: amazonia/classes/asg.py [hunk 4]
```diff
@@ -112,11 +118,68 @@ def create_launch_config(self, title, asg_config, network_config):
...
+            ScalingAdjustment=scaling_policy_config.scaling_adjustment,
+        ))
+
+        self.scaling_polcies.append(scaling_policy)
+
+        self.cw_alarms.append(self.template.add_resource(Alarm(
+            title=cf_name + 'Cwa',
+            AlarmActions=[Ref(scaling_policy)],
+            AlarmDescription=scaling_policy_config.description,
+            AlarmName=cf_name,
```
Comment (ga-autobots) [inline amazonia/classes/asg.py]: Check name

File: amazonia/classes/asg_config.py [hunk 5]
```diff
@@ -1,10 +1,17 @@
 #!/usr/bin/python3
 
+from amazonia.classes.util import detect_unencrypted_access_keys
+
+
+class InvalidAsgConfigError(Exception):
+    def __init__(self, value):
```
Comment (ga-autobots) [inline amazonia/classes/asg_config.py]: docstring

File: amazonia/classes/block_devices.py [hunk 6]
```diff
@@ -22,21 +22,21 @@ def __init__(self, title, block_devices_config):
...
-                    VolumeType=block_device['ebs_volume_type'])
-                if block_devices_config[n]['ebs_encrypted']:
-                    bdm.Ebs.Encrypted = block_device['ebs_encrypted']
-                if block_devices_config[n]['ebs_snapshot_id']:
-                    bdm.Ebs.SnapshotId = block_device['ebs_snapshot_id']
+                    VolumeSize=block_device.ebs_volume_size,
+                    VolumeType=block_device.ebs_volume_type)
+                if block_device.ebs_encrypted:
+                    bdm.Ebs.Encrypted = block_device.ebs_encrypted
+                if block_device.ebs_snapshot_id is not None:
```
Comment (ga-autobots) [inline amazonia/classes/block_devices.py]: remove is not none

File: amazonia/classes/elb.py [hunk 7]
```diff
@@ -58,30 +57,34 @@ def __init__(self, title, template, network_config, elb_config):
...
+            self.create_r53_record(network_config.stack_hosted_zone_name)
         else:
             self.template.add_output(Output(
                 self.trop_elb.title,
                 Description='URL of the {0} ELB'.format(self.title),
                 Value=Join('', ['http://', GetAtt(self.trop_elb, 'DNSName')])
             ))
+
+    def create_r53_record(self, hosted_zone_name):
+        self.elb_r53 = self.template.add_resource(route53.RecordSetGroup(
```
Comment (ga-autobots) [inline amazonia/classes/elb.py]: docstring

File: amazonia/classes/stack.py [hunk 8]
```diff
@@ -187,64 +180,33 @@ def __init__(self, stack_title, code_deploy_service_role, keypair, availability_
...
             for dependency in dependencies:
                 self.units[unit_name].add_unit_flow(self.units[dependency])
 
+    def add_units(self, unit_list, unit_constructor):
+        for unit in unit_list:  # type: dict
+            unit_title = unit['unit_title']
+            if unit_title in self.units:
+                raise DuplicateUnitNameError("Error: unit name '{0}' has already been specified, "
+                                             'it must be unique.'.format(unit_title))
+            unit['unit_title'] = unit_title
```
Comment (ga-autobots) [inline amazonia/classes/stack.py]: remove unit title

File: amazonia/classes/util.py [hunk 9]
```diff
@@ -1,11 +1,17 @@
...
 import yaml
 
 logging.basicConfig(
     level=logging.INFO
 )
 
 
+def get_cf_friendly_name(object_name):
+    return re.sub(r'\W+', '', object_name)
+
```
Comment (ga-autobots) [inline amazonia/classes/util.py]: camel case

File: amazonia/classes/util.py [hunk 10]
```diff
@@ -1,11 +1,17 @@
...
 import yaml
 
 logging.basicConfig(
     level=logging.INFO
 )
 
 
+def get_cf_friendly_name(object_name):
+    return re.sub(r'\W+', '', object_name)
```
Comment (ga-autobots) [inline amazonia/classes/util.py]: docstring

File: amazonia/classes/util.py [hunk 11]
```diff
@@ -14,3 +20,24 @@ def read_yaml(user_yaml):
...
+    if aws_access_id.search(userdata):
+        raise InsecureVariableError('Error: unencrypted {0} was found in your userdata, please remove or encrypt.'
+                                    .format('AWS access ID'))
+    elif aws_secret_key.search(userdata):
+        raise InsecureVariableError('Error: unencrypted {0} was found in your userdata, please remove or encrypt.'
+                                    .format('AWS secret Key'))
+
+
+class InsecureVariableError(Exception):
+    def __init__(self, value):
```
Comment (ga-autobots) [inline amazonia/classes/util.py]: docstring

File: amazonia/classes/yaml.py [hunk 12]
```diff
@@ -119,82 +36,38 @@ def __init__(self, user_stack_data, default_data):
...
+            value = user_values.get(current_key, default_values[current_key])
+        return value
+
+    def get_complex_params(self, current_key, nested_user_values, complex_key_list, default_values):
+        complex_params = {}
+        for complex_key in complex_key_list:
+            if complex_key in default_values:
+                nested_default_values = default_values
+            else:
+                nested_default_values = default_values[current_key]
```
Comment (ga-autobots) [inline amazonia/classes/yaml.py]: check this logic

File: amazonia/classes/util.py [hunk 13]
```diff
@@ -14,3 +25,27 @@ def read_yaml(user_yaml):
...
+        raise InsecureVariableError('Error: unencrypted {0} was found in your userdata, please remove or encrypt.'
+                                    .format('AWS access ID'))
+    elif aws_secret_key.search(userdata):
+        raise InsecureVariableError('Error: unencrypted {0} was found in your userdata, please remove or encrypt.'
+                                    .format('AWS secret Key'))
+
+
+class InsecureVariableError(Exception):
+    """
+    Exception if aws access ud ir aws secret key is detected in userdata
```
Comment (fuddstar) [inline amazonia/classes/util.py]: typo access id or...?

File: amazonia/classes/stack.py (hunk 8 above)
Comment (snayler0) [inline amazonia/classes/stack.py]: Unit_title still here. Did we find it was still needed?

**Comment by fuddstar** [comment]
Approved

**Comment by snayler0** [comment]
Approved

**Pull request merged**